docs/Options.md
# Options

You can set these options by using:

```
/set plugins.var.python.slack.option_name value
```

You can also see all the options and set them them interactively by running `/fset slack`.

Note that the default value will be shown as an empty string in WeeChat.
The actual default values are listed below.

Most options require that you reload the script with `/python reload
slack` after changing it to take effect.

## Available options:

### auto_open_threads

**Default:** `false`

**Description:** Automatically open threads when mentioned or in response to own messages.

### background_load_all_history

**Default:** `true`

**Description:** Load the history for all channels in the background when the script is loaded, rather than waiting until the buffer is switched to. You can set this to false if you experience performance issues, however that causes some loss of functionality, see known issues in the readme.

### channel_name_typing_indicator

**Default:** `true`

**Description:** Change the prefix of a channel from # to > when someone is typing in it. Note that this will (temporarily) affect the sort order if you sort buffers by name rather than by number.

### channel_section_buffer

**Default:** `false`

**Description**: Enable to create virtual buffers for sorting channels using channel sections. Use autosort.py for better sorting.

### color_buflist_muted_channels

**Default:** `darkgray`

**Description:** Color to use for muted channels in the buflist

### color_deleted

**Default:** `red`

**Description:** Color to use for deleted messages and files.

### color_edited_suffix

**Default:** `095`

**Description:** Color to use for (edited) suffix on messages that have been edited.

### color_reaction_suffix

**Default:** `darkgray`

**Description:** Color to use for the [:wave:(@user)] suffix on messages that have reactions attached to them.

### color_reaction_suffix_added_by_you

**Default:** `blue`

**Description:** Color to use for reactions that you have added.

### color_thread_suffix

**Default:** `lightcyan`

**Description:** Color to use for the [thread: XXX] suffix on messages that have threads attached to them. The special value "multiple" can be used to use a different color for each thread.

### color_typing_notice

**Default:** `yellow`

**Description:** Color to use for the typing notice.

### colorize_attachments

**Default:** `prefix`

**Description:** Whether to colorize attachment lines. Values: "prefix": Only colorize the prefix, "all": Colorize the whole line, "none": Don't colorize.

### colorize_private_chats

**Default:** `false`

**Description:** Whether to use nick-colors in DM windows.

### debug_level

**Default:** `3`

**Description:** Show only this level of debug info (or higher) when debug_mode is on. Lower levels -> more messages.

### debug_mode

**Default:** `false`

**Description:** Open a dedicated buffer for debug messages and start logging to it. How verbose the logging is depends on log_level.

### distracting_channels

**Default:** ``

**Description:** List of channels to hide.

### external_user_suffix

**Default:** `*`

**Description:** The suffix appended to nicks to indicate external users.

### files_download_location

**Default:** ``

**Description:** If set, file attachments will be automatically downloaded to this location. "%h" will be replaced by WeeChat home, "~/.weechat" by default. Requires WeeChat 2.2 or newer.

### group_name_prefix

**Default:** `&`

**Description:** The prefix of buffer names for groups (private channels).

### history_fetch_count

**Default:** `200`

**Description:** The number of messages to fetch for each channel when fetching history, between 1 and 1000.

### link_previews

**Default:** `true`

**Description:** Show previews of website content linked by teammates.

### map_underline_to

**Default:** `_`

**Description:** When sending underlined text to slack, use this formatting character for it. The default ("_") sends it as italics. Use "*" to send bold instead.

### muted_channels_activity

**Default:** `personal_highlights`

**Description:** Control which activity you see from muted channels, either none, personal_highlights, all_highlights or all. none: Don't show any activity. personal_highlights: Only show personal highlights, i.e. not @channel and @here. all_highlights: Show all highlights, but not other messages. all: Show all activity, like other channels.

### never_away

**Default:** `false`

**Description:** Poke Slack every five minutes so that it never marks you "away".

### notify_subscribed_threads

**Default:** `auto`

**Description:** Control if you want to see a notification in the team buffer when a thread you're subscribed to receives a new message, either auto, true or false. auto means that you only get a notification if auto_open_threads and thread_messages_in_channel both are false. Defaults to auto.

### notify_usergroup_handle_updated

**Default:** `false`

**Description:** Control if you want to see a notification in the team buffer when ausergroup's handle has changed, either true or false.

### record_events

**Default:** `false`

**Description:** Log all traffic from Slack to disk as JSON.

### render_bold_as

**Default:** `bold`

**Description:** When receiving bold text from Slack, render it as this in WeeChat.

### render_emoji_as_string

**Default:** `false`

**Description:** Render emojis as :emoji_name: instead of emoji characters. Enable this if your terminal doesn't support emojis, or set to 'both' if you want to see both renderings. Note that even though this is disabled by default, you need to place https://github.com/wee-slack/wee-slack/blob/master/weemoji.json in your WeeChat directory to enable rendering emojis as emoji characters.

### render_italic_as

**Default:** `italic`

**Description:** When receiving bold text from Slack, render it as this in WeeChat. If your terminal lacks italic support, consider using "underline" instead.

### send_typing_notice

**Default:** `true`

**Description:** Alert Slack users when you are typing a message in the input bar (Requires reload)

### server_aliases

**Default:** ``

**Description:** A comma separated list of `subdomain:alias` pairs. The alias will be used instead of the actual name of the slack (in buffer names, logging, etc). E.g `work:no_fun_allowed` would make your work slack show up as `no_fun_allowed` rather than `work.slack.com`.

### shared_name_prefix

**Default:** `%`

**Description:** The prefix of buffer names for shared channels.

### short_buffer_names

**Default:** `false`

**Description:** Use `foo.#channel` rather than `foo.slack.com.#channel` as the internal name for Slack buffers.

### show_buflist_presence

**Default:** `true`

**Description:** Display a `+` character in the buffer list for present users.

### show_reaction_nicks

**Default:** `false`

**Description:** Display the name of the reacting user(s) alongside each reactji.

### slack_api_token

**Default:** `INSERT VALID KEY HERE!`

**Description:** List of Slack API tokens, one per Slack instance you want to connect to; see the README for details on how to get these (note: content is evaluated, see /help eval).

### slack_timeout

**Default:** `20000`

**Description:** How long (ms) to wait when communicating with Slack.

### switch_buffer_on_join

**Default:** `true`

**Description:** When /joining a channel, automatically switch to it as well.

### thread_broadcast_prefix

**Default:** `+ `

**Description:** Prefix to distinguish thread messages that were also sent to the channel, when thread_messages_in_channel is enabled.

### thread_messages_in_channel

**Default:** `false`

**Description:** When enabled shows thread messages in the parent channel.

### unfurl_auto_link_display

**Default:** `both`

**Description:** When displaying ("unfurling") links to channels/users/etc, determine what is displayed when the text matches the url without the protocol. This happens when Slack automatically creates links, e.g. from words separated by dots or email addresses. Set it to "text" to only display the text written by the user, "url" to only display the url or "both" (the default) to display both.

### unfurl_ignore_alt_text

**Default:** `false`

**Description:** When displaying ("unfurling") links to channels/users/etc, ignore the "alt text" present in the message and instead use the canonical name of the thing being linked to.

### unhide_buffers_with_activity

**Default:** `false`

**Description:** When activity occurs on a buffer, unhide it even if it was previously hidden (whether by the user or by the distracting_channels setting).

### use_full_names

**Default:** `false`

**Description:** Use full names as the nicks for all users. When this is false (the default), display names will be used if set, with a fallback to the full name if display name is not set.



CHANGELOG.md
# Changelog

## 2.11.0 (2024-10-09)

### Slack API changes

- Prevent crash on startup after API response from `users.prefs.get` was changed (fixes #930).
- Fix muted channels not being updated after API change.

### WeeChat compatibility

- Set buffer property `input_prompt` to the users nick.

### Features

- Make logging "Channel Created" optional (PR #926).
- Support filtering users list based on regular expression (PR #929).
- Add option to see the username as the nick (PR #944).

### Bug fixes

- Fix IM title not being updated on user status change.

## 2.10.2 (2024-02-18)

Note: This will most likely be the last version before version 3.0.0, which is a mostly complete rewrite of the script with some breaking changes. Notably the configuration options have completely changed to be more in line with how the IRC plugin does it with a separate section for each Slack workspace. There will be some new features, but some existing features might also be dropped. Several bugs have been fixed, but of course new ones may have been introduced too.

### Bug fixes

- Fix rendering of text styles for certain messages (regression from version v2.10.0).
- Only open subscribed threads with `auto_open_threads` enabled (fixes #803, fixes #830).
- Mark conversation/thread as read when `/buffer set unread` is run.
- Handle millisecond-level timestamps in attachments (PR #905).
- Handle Firefox containers in `extract_token_from_browser.py` (PR #909).
- Properly fix rendering of `<` and `>` (fixes #908). The fix in v2.10.1 only fixed it for normal text, not e.g. code blocks and attachments.
- Support color rich text elements (fixes #912).
- Correctly show reactions with more than 50 users.
- Make /thread work on thread replies.
- Add no-secretstorage option for `extract_token_from_browser.py`.

## 2.10.1 (2023-09-22)

### Slack API changes

- Fix the connection to the websocket when using session (`xoxc-`) tokens. These connections started failing after a change in the Slack API which made wee-slack unable to run at all (fixes #901).

### Bug fixes

- Fix being invited to a private channel not working.
- Add join messages for your own user to hotlist (if unread).
- Prevent errors being printed in core buffer on events for unknown channels.
- Fix rendering of `<` and `>`. A regression in 2.10.0 made these be removed.
- Fix rendering of huddles. A regression in 2.10.0 made the text for huddle messages duplicated.
- Fix rendering of emojis when `render_emoji_as_string` is set to `both`. A regression in 2.10.0 made emojis show twice with this option (PR #902).
- Print an error message when running the `/reply` command without any arguments (fixes #900).

## 2.10.0 (2023-08-24)

### WeeChat compatibility

- Support multiline rendering in WeeChat >= 4.0.0.

### Features

- Display link to join huddle when started (PR #885).
- Support Chromium, Chrome and Firefox snap in `extract_token_from_browser.py` (PR #884).
- Detect default Firefox profile in `extract_token_from_browser.py` (PR #887).
- Support specifying profile to use in `extract_token_from_browser.py` (PR #884, PR #887).
- Read new local storage format for Firefox in `extract_token_from_browser.py` (PR #887).
- Automatically enable emoji/usergroups completion.
- Add info to get contents of a message (fixes #889).
- Filter channel list based on regular expression (PR #896).

### Bug fixes

- Support `/msg *` in thread buffers (fixes #888).
- Improve rendering of messages by using the `rich_text` block (fixes #354, fixes #550, fixes #893).

### Slack API changes

- Fix showing origin channel in previews of Slack messages when a message is linked after it was broken by an API change.

### Other changes

- Include a space between nicks in reaction string.
- Update Dockerfile to use XDG directories (PR #894).

## 2.9.1 (2022-10-30)

### WeeChat compatibility

- Update WeeChat version check to require >= 2.2. The reason for requiring this is that a feature of WeeChat 2.2 (specifying HTTP cookies) is necessary for adding support for the new type of session tokens added in version 2.9.0, so this should have been updated in that version.
- Update usage of some deprecated WeeChat API methods.

### Bug fixes

- Open open IMs and MPIMs on start, not just unread (fixes #875).
- Fix old type MPIMs not being marked as unread on start.
- Evaluate value of the `weechat.look.buffer_time_format` option (fixes #871).
- Fix link previews from apps not being recognized (fixes #834).
- Show useful link for Slack posts.

slack/config.py
from __future__ import annotations

from typing import TYPE_CHECKING, Callable, Optional, Tuple

import weechat

from slack.log import print_error
from slack.shared import shared
from slack.slack_conversation import invalidate_nicklists, update_buffer_props
from slack.slack_workspace import SlackWorkspace, workspace_get_buffer_to_merge_with
from slack.util import get_callback_name
from slack.weechat_config import (
    WeeChatColor,
    WeeChatConfig,
    WeeChatOption,
    WeeChatOptionType,
    WeeChatSection,
)

if TYPE_CHECKING:
    from typing_extensions import Literal


class SlackConfigSectionColor:
    def __init__(self, weechat_config: WeeChatConfig):
        self._section = WeeChatSection(weechat_config, "color")

        self.buflist_muted_conversation = WeeChatOption(
            self._section,
            "buflist_muted_conversation",
            "text color for muted conversations in the buflist",
            WeeChatColor("darkgray"),
            callback_change=self.config_change_buflist_muted_conversation_cb,
        )

        self.channel_mention = WeeChatOption(
            self._section,
            "channel_mention",
            "text color for mentioned channel names in the chat",
            WeeChatColor("blue"),
        )

        self.deleted_message = WeeChatOption(
            self._section,
            "deleted_message",
            "text color for a deleted message",
            WeeChatColor("red"),
        )

        self.disconnected = WeeChatOption(
            self._section,
            "disconnected",
            "text color for the disconnected text",
            WeeChatColor("red"),
        )

        self.edited_message_suffix = WeeChatOption(
            self._section,
            "edited_message_suffix",
            "text color for the suffix after an edited message",
            WeeChatColor("095"),
        )

        self.loading = WeeChatOption(
            self._section,
            "loading",
            "text color for the loading text",
            WeeChatColor("yellow"),
        )

        self.message_join = WeeChatOption(
            self._section,
            "message_join",
            "color for text in join messages",
            WeeChatColor("green"),
            parent_option="irc.color.message_join",
        )

        self.message_quit = WeeChatOption(
            self._section,
            "message_quit",
            "color for text in part messages",
            WeeChatColor("red"),
            parent_option="irc.color.message_quit",
        )

        self.reaction_suffix = WeeChatOption(
            self._section,
            "reaction_suffix",
            "color for the reactions after a message",
            WeeChatColor("darkgray"),
        )

        self.reaction_self_suffix = WeeChatOption(
            self._section,
            "reaction_self_suffix",
            "color for the reactions after a message, for reactions you have added",
            WeeChatColor("blue"),
        )

        self.render_error = WeeChatOption(
            self._section,
            "render_error",
            "color for displaying rendering errors in a message",
            WeeChatColor("red"),
        )

        self.search_line_marked_bg = WeeChatOption(
            self._section,
            "search_line_marked_bg",
            "background color for a marked line in search buffers",
            WeeChatColor("17"),
        )

        self.search_line_selected_bg = WeeChatOption(
            self._section,
            "search_line_selected_bg",
            "background color for the selected line in search buffers",
            WeeChatColor("24"),
        )

        self.search_marked = WeeChatOption(
            self._section,
            "search_marked",
            "color for mark indicator in search buffers",
            WeeChatColor("brown"),
        )

        self.search_marked_selected = WeeChatOption(
            self._section,
            "search_marked_selected",
            "color for mark indicator on the selected line in search buffers",
            WeeChatColor("yellow"),
        )

        self.user_mention = WeeChatOption(
            self._section,
            "user_mention",
            "text color for mentioned user names in the chat",
            WeeChatColor("blue"),
        )

        self.user_mention_nick_color = WeeChatOption(
            self._section,
            "user_mention_nick_color",
            "",
            False,
            callback_change=self.config_change_user_mention_nick_color_cb,
        )

        self.usergroup_mention = WeeChatOption(
            self._section,
            "usergroup_mention",
            "text color for mentioned user group names in the chat",
            WeeChatColor("blue"),
        )

    def config_change_buflist_muted_conversation_cb(
        self, option: WeeChatOption[WeeChatOptionType], parent_changed: bool
    ):
        update_buffer_props()

    def config_change_user_mention_nick_color_cb(
        self, option: WeeChatOption[WeeChatOptionType], parent_changed: bool
    ):
        self.user_mention.enabled = not option.value


class SlackConfigSectionLook:
    def __init__(self, weechat_config: WeeChatConfig):
        self._section = WeeChatSection(weechat_config, "look")

        self.bot_user_suffix = WeeChatOption(
            self._section,
            "bot_user_suffix",
            "the suffix appended to nicks to indicate a bot",
            " :]",
        )

        self.thread_broadcast_prefix = WeeChatOption(
            self._section,
            "thread_broadcast_prefix",
            "prefix to distinguish thread messages that were also sent to the channel, when display_thread_replies_in_channel is enabled",
            "+",
        )

        self.color_nicks_in_nicklist = WeeChatOption(
            self._section,
            "color_nicks_in_nicklist",
            "use nick color in nicklist",
            False,
            parent_option="irc.look.color_nicks_in_nicklist",
            callback_change=self.config_change_color_nicks_in_nicklist_cb,
        )

        self.color_message_attachments: WeeChatOption[
            Literal["prefix", "all", "none"]
        ] = WeeChatOption(
            self._section,
            "color_message_attachments",
            "colorize attachments in a message: prefix = only colorize the prefix, all = colorize the whole line, none = don't colorize",
            "prefix",
            string_values=("prefix", "all", "none"),
        )

        self.display_link_previews: WeeChatOption[
            Literal["always", "only_internal", "never"]
        ] = WeeChatOption(
            self._section,
            "display_link_previews",
            "display previews of URLs in messages: always = always display, only_internal = only display for URLs to messages in the workspace, never = never display",
            "always",
            string_values=("always", "only_internal", "never"),
        )

        self.display_reaction_nicks = WeeChatOption(
            self._section,
            "display_reaction_nicks",
            "display the name of the reacting user(s) after each reaction; can be overridden per buffer with the buffer localvar display_reaction_nicks",
            False,
        )

        self.display_thread_replies_in_channel = WeeChatOption(
            self._section,
            "display_thread_replies_in_channel",
            "display thread replies in the parent channel; can be overridden per buffer with the buffer localvar display_thread_replies_in_channel; note that it only takes effect for new messages; note that due to limitations in the Slack API, on load only thread messages for parents that are in the buffer and thread messages in subscribed threads will be displayed (but all thread messages received while connected will be displayed)",
            False,
        )

        self.external_user_suffix = WeeChatOption(
            self._section,
            "external_user_suffix",
            "the suffix appended to nicks to indicate external users",
            "*",
        )

        self.leave_channel_on_buffer_close = WeeChatOption(
            self._section,
            "leave_channel_on_buffer_close",
            "leave channel when a buffer is closed",
            True,
        )

        self.muted_conversations_notify: WeeChatOption[
            Literal["none", "personal_highlights", "all_highlights", "all"]
        ] = WeeChatOption(
            self._section,
            "muted_conversations_notify",
            "notify level to set for messages in muted conversations; none: don't notify for any messages; personal_highlights: only notify for personal highlights, i.e. not @channel and @here; all_highlights: notify for all highlights, but not other messages; all: notify for all messages, like other channels; note that this doesn't affect messages in threads you are subscribed to or in open thread buffers, those will always notify",
            "personal_highlights",
            string_values=("none", "personal_highlights", "all_highlights", "all"),
        )

        self.notify_subscribed_threads: WeeChatOption[
            Literal["auto", "unless_thread_buffer", "always", "never"]
        ] = WeeChatOption(
            self._section,
            "notify_subscribed_threads",
            "send a message to the workspace buffer to notify you of new messages in threads you are subscribed to: auto = only notify if the thread buffer is not open and display_thread_replies_in_channel for the channel is false, unless_thread_buffer = only notify if the thread buffer is not open, always = always notify, never = never notify",
            "auto",
            string_values=("auto", "unless_thread_buffer", "always", "never"),
        )

        self.part_closes_buffer = WeeChatOption(
            self._section,
            "part_closes_buffer",
            "close buffer when /slack part is issued on a channel",
            False,
            parent_option="irc.look.part_closes_buffer",
        )

        self.render_emoji_as: WeeChatOption[Literal["emoji", "name", "both"]] = (
            WeeChatOption(
                self._section,
                "render_emoji_as",
                "show emojis as: emoji = the emoji unicode character, name = the emoji name, both = both the emoji name and the emoji character",
                "emoji",
                string_values=("emoji", "name", "both"),
            )
        )

        self.render_url_as = WeeChatOption(
            self._section,
            "render_url_as",
            "format to render URLs (note: content is evaluated, see /help eval; ${url} is replaced by the URL link and ${text} is replaced by the URL text); the default format renders only the URL if the text is empty or is contained in the URL, otherwise it renders the text (underlined) first and then the URL in parentheses",
            "${if: ${text} == || ${url} =- ${text} ?${url}:${color:underline}${text}${color:-underline} (${url})}",
        )

        self.replace_space_in_nicks_with = WeeChatOption(
            self._section,
            "replace_space_in_nicks_with",
            "",
            "",
        )

        self.workspace_buffer: WeeChatOption[
            Literal["merge_with_core", "merge_without_core", "independent"]
        ] = WeeChatOption(
            self._section,
            "workspace_buffer",
            "merge workspace buffers; this option has no effect if a layout is saved and is conflicting with this value (see /help layout)",
            "merge_with_core",
            string_values=("merge_with_core", "merge_without_core", "independent"),
            parent_option="irc.look.server_buffer",
            callback_change=self.config_change_workspace_buffer_cb,
        )

        self.typing_status_nicks = WeeChatOption(
            self._section,
            "typing_status_nicks",
            'display nicks typing on the channel in bar item "typing" (option typing.look.enabled_nicks must be enabled)',
            True,
        )

        self.typing_status_self = WeeChatOption(
            self._section,
            "typing_status_self",
            "send self typing status to channels so that other users see when you are typing a message (option typing.look.enabled_self must be enabled)",
            True,
        )

        weechat.hook_config(
            "weechat.look.nick_color_*",
            get_callback_name(self.config_change_nick_colors_cb),
            "",
        )
        weechat.hook_config(
            "weechat.color.chat_nick_colors",
            get_callback_name(self.config_change_nick_colors_cb),
            "",
        )

    def config_change_color_nicks_in_nicklist_cb(
        self, option: WeeChatOption[WeeChatOptionType], parent_changed: bool
    ):
        invalidate_nicklists()

    def config_change_workspace_buffer_cb(
        self, option: WeeChatOption[WeeChatOptionType], parent_changed: bool
    ):
        for workspace in shared.workspaces.values():
            if workspace.buffer_pointer:
                weechat.buffer_unmerge(workspace.buffer_pointer, -1)

        buffer_to_merge_with = workspace_get_buffer_to_merge_with()
        if buffer_to_merge_with:
            for workspace in shared.workspaces.values():
                if (
                    workspace.buffer_pointer
                    and workspace.buffer_pointer != buffer_to_merge_with
                ):
                    weechat.buffer_merge(workspace.buffer_pointer, buffer_to_merge_with)

    def config_change_nick_colors_cb(self, data: str, option: str, value: str):
        invalidate_nicklists()
        return weechat.WEECHAT_RC_OK


class SlackConfigSectionWorkspace:
    def __init__(
        self,
        section: WeeChatSection,
        workspace_name: Optional[str],
        parent_config: Optional[SlackConfigSectionWorkspace],
    ):
        self._section = section
        self._workspace_name = workspace_name
        self._parent_config = parent_config

        self.api_token = self._create_option(
            "api_token",
            "The token (note: content is evaluated, see /help eval; workspace options are evaluated with ${workspace} replaced by the workspace name)",
            "",
            evaluate_func=self._evaluate_with_workspace_name,
        )

        self.api_cookies = self._create_option(
            "api_cookies",
            "The cookies (note: content is evaluated, see /help eval; workspace options are evaluated with ${workspace} replaced by the workspace name)",
            "",
            evaluate_func=self._evaluate_with_workspace_name,
        )

        self.auto_open_threads = self._create_option(
            "auto_open_threads",
            "automatically open thread buffers; see the other options starting with auto_open_threads for which threads to open; can be overridden per buffer with the buffer localvar auto_open_threads",
            False,
        )

        self.auto_open_threads_only_if_replies_not_in_channel = self._create_option(
            "auto_open_threads_only_if_replies_not_in_channel",
            "limit automatically opening threads to only threads in conversations where display_thread_replies_in_channel is disabled; can be overridden per buffer with the buffer localvar auto_open_threads_only_if_replies_not_in_channel",
            True,
        )

        self.auto_open_threads_only_subscribed = self._create_option(
            "auto_open_threads_only_subscribed",
            "limit automatically opening threads to only subscribed threads; note that only subscribed threads have a read status on the server, so on script load all messages in unsubscribed threads will be considered read; can be overridden per buffer with the buffer localvar auto_open_threads_only_subscribed",
            True,
        )

        self.auto_open_threads_only_unread = self._create_option(
            "auto_open_threads_only_unread",
            "limit automatically opening threads to only unread threads; note that only subscribed threads have a read status on the server, so this option only applies to subscribed threads; can be overridden per buffer with the buffer localvar auto_open_threads_only_unread",
            True,
        )

        self.autoconnect = self._create_option(
            "autoconnect",
            "automatically connect to workspace when WeeChat is starting",
            False,
        )

        self.keep_active: WeeChatOption[Literal["on_activity", "always"]] = (
            self._create_option(
                "keep_active",
                "keep your presence set to active: on_activity = set active when you interact with WeeChat (Slack sets you away after 30 minutes of inactivity), always = remain active as long as you're connected to the workspace",
                "on_activity",
                string_values=("on_activity", "always"),
            )
        )

        self.network_timeout = self._create_option(
            "network_timeout",
            "timeout (in seconds) for network requests",
            30,
        )

        self.nick_source: WeeChatOption[
            Literal["display_name", "real_name", "username"]
        ] = self._create_option(
            "nick_source",
            "property from the user profile to use as the nick",
            "display_name",
            string_values=("display_name", "real_name", "username"),
        )

    def _evaluate_with_workspace_name(self, value: str) -> str:
        return weechat.string_eval_expression(
            value, {}, {"workspace": self._workspace_name or ""}, {}
        )

    def _create_option(
        self,
        name: str,
        description: str,
        default_value: WeeChatOptionType,
        min_value: Optional[int] = None,
        max_value: Optional[int] = None,
        string_values: Tuple[WeeChatOptionType, ...] = (),
        evaluate_func: Optional[
            Callable[[WeeChatOptionType], WeeChatOptionType]
        ] = None,
    ) -> WeeChatOption[WeeChatOptionType]:
        if self._workspace_name:
            option_name = f"{self._workspace_name}.{name}"
        else:
            option_name = name

        if self._parent_config:
            parent_option = getattr(self._parent_config, name, None)
        else:
            parent_option = None

        return WeeChatOption(
            self._section,
            option_name,
            description,
            default_value,
            min_value,
            max_value,
            string_values,
            parent_option,
            evaluate_func=evaluate_func,
        )


def config_section_workspace_read_cb(
    data: str, config_file: str, section: str, option_name: str, value: Optional[str]
) -> int:
    option_split = option_name.split(".", maxsplit=1)
    if len(option_split) < 2:
        return weechat.WEECHAT_CONFIG_OPTION_SET_ERROR
    workspace_name, name = option_split
    if not workspace_name or not name:
        return weechat.WEECHAT_CONFIG_OPTION_SET_ERROR

    if workspace_name not in shared.workspaces:
        shared.workspaces[workspace_name] = SlackWorkspace(workspace_name)

    option = getattr(shared.workspaces[workspace_name].config, name, None)
    if option is None:
        return weechat.WEECHAT_CONFIG_OPTION_SET_OPTION_NOT_FOUND
    if not isinstance(option, WeeChatOption):
        return weechat.WEECHAT_CONFIG_OPTION_SET_ERROR

    if value is None or (
        shared.weechat_version < 0x03080000
        and value == ""
        and option.weechat_type != "string"
    ):
        rc = option.value_set_null()
    else:
        rc = option.value_set_as_str(value)
    if rc == weechat.WEECHAT_CONFIG_OPTION_SET_ERROR:
        print_error(f'error creating workspace option "{option_name}"')
    return rc


def config_section_workspace_write_for_old_weechat_cb(
    data: str, config_file: str, section_name: str
) -> int:
    if not weechat.config_write_line(config_file, section_name, ""):
        return weechat.WEECHAT_CONFIG_WRITE_ERROR

    for workspace in shared.workspaces.values():
        for option in vars(workspace.config).values():
            if (
                isinstance(option, WeeChatOption)
                and option.pointer is not None
                and (
                    option.weechat_type != "string"
                    or not weechat.config_option_is_null(option.pointer)
                )
            ):
                if not weechat.config_write_option(config_file, option.pointer):
                    return weechat.WEECHAT_CONFIG_WRITE_ERROR

    return weechat.WEECHAT_CONFIG_WRITE_OK


class SlackConfig:
    def __init__(self):
        self.weechat_config = WeeChatConfig("slack")
        self.color = SlackConfigSectionColor(self.weechat_config)
        self.look = SlackConfigSectionLook(self.weechat_config)
        self._section_workspace_default = WeeChatSection(
            self.weechat_config, "workspace_default"
        )
        # WeeChat < 3.8 sends null as an empty string to callback_read, so in
        # order to distinguish them, don't write the null values to the config
        # See https://github.com/weechat/weechat/pull/1843
        callback_write = (
            get_callback_name(config_section_workspace_write_for_old_weechat_cb)
            if shared.weechat_version < 0x03080000
            else ""
        )
        self._section_workspace = WeeChatSection(
            self.weechat_config,
            "workspace",
            callback_read=get_callback_name(config_section_workspace_read_cb),
            callback_write=callback_write,
        )
        self._workspace_default = SlackConfigSectionWorkspace(
            self._section_workspace_default, None, None
        )

    def config_read(self):
        weechat.config_read(self.weechat_config.pointer)

    def create_workspace_config(self, workspace_name: str):
        if workspace_name in shared.workspaces:
            raise Exception(
                f"Failed to create workspace config, already exists: {workspace_name}"
            )
        return SlackConfigSectionWorkspace(
            self._section_workspace, workspace_name, self._workspace_default
        )
